File: 01.python/Security_Hotspot/Security37.py
###### Использование незашифрованных файловых систем EFS важно для безопасности.

Точка доступа безопасности
Главный

     Доступно с 19 декабря 2023 г.
     СонарКуб (Python)
     Постоянная/проблема: 10 мин.

Amazon Elastic File System (EFS) — это бессерверная файловая система, не требующая выделения ресурсов или управления хранилищем. Сохраненные файлы могут быть автоматически зашифрованы службой. В случае, если злоумышленники получают физический доступ к носителю данных или иным образом сливают сообщение, они не могут получить доступ к данным.

###### Спросите себя, есть ли

     Файловая система содержит конфиденциальные данные, утечка которых может причинить вред.
     К службе хранения данных в зашифрованном виде предъявляются требования соответствия.

Существует риск, если вы ответили утвердительно на любой из этих вопросов.
Пример конфиденциального кода

For aws_cdk.aws_efs.FileSystem and aws_cdk.aws_efs.CfnFileSystem:

from aws_cdk import (
    aws_efs as efs
)

efs.FileSystem(
    self,
    "example",
    encrypted=False  # Sensitive
)




### Рекомендуемые методы безопасного кодирования

Рекомендуется шифровать файловые системы EFS, содержащие конфиденциальную информацию. Шифрование и дешифрование прозрачно выполняются EFS, поэтому дальнейшие изменения приложения не требуются.
Соответствующее решение


For aws_cdk.aws_efs.FileSystem and aws_cdk.aws_efs.CfnFileSystem:

from aws_cdk import (
    aws_efs as efs
)

efs.FileSystem(
    self,
    "example",
    encrypted=True
)

See

    OWASP Top 10 2021 Category A2 - Cryptographic Failures
    OWASP Top 10 2021 Category A4 - Insecure Design
    OWASP Top 10 2021 Category A5 - Security Misconfiguration
    AWS Documentation - Data encryption in Amazon EFS
    OWASP Top 10 2017 Category A3 - Sensitive Data Exposure
    OWASP Top 10 2017 Category A6 - Security Misconfiguration
    MITRE, CWE-311 - Missing Encryption of Sensitive Data

File: 00.java/01.Security_Hotspot/Security37.java
###### Using weak hashing algorithms is security-sensitive

Security Hotspot
Critical

    Available SinceDec 19, 2023
    SonarQube (Java)

Cryptographic hash algorithms such as MD2, MD4, MD5, MD6, HAVAL-128, HMAC-MD5, DSA (which uses SHA-1), RIPEMD, RIPEMD-128, RIPEMD-160, HMACRIPEMD160 and SHA-1 are no longer considered secure, because it is possible to have collisions (little computational effort is enough to find two or more different inputs that produce the same hash).


###### Ask Yourself Whether

The hashed value is used in a security context like:

    User-password storage.
    Security token generation (used to confirm e-mail when registering on a website, reset password, etc …​).
    To compute some message integrity.

There is a risk if you answered yes to any of those questions.
Sensitive Code Example

MessageDigest md1 = MessageDigest.getInstance("SHA");  // Sensitive:  SHA is not a standard name, for most security providers it's an alias of SHA-1
MessageDigest md2 = MessageDigest.getInstance("SHA1");  // Sensitive






######## Recommended Secure Coding Practices

Safer alternatives, such as SHA-256, SHA-512, SHA-3 are recommended, and for password hashing, it’s even better to use algorithms that do not compute too "quickly", like bcrypt, scrypt, argon2 or pbkdf2 because it slows down brute force attacks.
Compliant Solution

MessageDigest md1 = MessageDigest.getInstance("SHA-512"); // Compliant

See

    OWASP Top 10 2021 Category A2 - Cryptographic Failures
    OWASP Top 10 2017 Category A3 - Sensitive Data Exposure
    OWASP Top 10 2017 Category A6 - Security Misconfiguration
    Mobile AppSec Verification Standard - Cryptography Requirements
    OWASP Mobile Top 10 2016 Category M5 - Insufficient Cryptography
    MITRE, CWE-1240 - Use of a Risky Cryptographic Primitive
    SANS Top 25 - Porous Defenses




